lib/yarp/desugar_visitor.rb

```diff
@@ -0,0 +1,267 @@
+# frozen_string_literal: true
+
+module YARP
+  class DesugarVisitor < MutationVisitor
+    # @@foo &&= bar
+    #
+    # becomes
+    #
+    # @@foo && @@foo = bar
+    def visit_class_variable_and_write_node(node)
+      AndNode.new(
+        ClassVariableReadNode.new(node.name_loc),
+        ClassVariableWriteNode.new(node.name_loc, node.value, node.operator_loc, node.location),
+        node.operator_loc,
+        node.location
+      )
+    end
+
+    # @@foo ||= bar
+    #
+    # becomes
+    #
+    # @@foo || @@foo = bar
+    def visit_class_variable_or_write_node(node)
+      OrNode.new(
+        ClassVariableReadNode.new(node.name_loc),
+        ClassVariableWriteNode.new(node.name_loc, node.value, node.operator_loc, node.location),
+        node.operator_loc,
+        node.location
+      )
+    end
+
+    # @@foo += bar
+    #
+    # becomes
+    #
+    # @@foo = @@foo + bar
+    def visit_class_variable_operator_write_node(node)
+      desugar_operator_write_node(node, ClassVariableWriteNode, ClassVariableReadNode)
+    end
+
+    # Foo &&= bar
+    #
+    # becomes
+    #
+    # Foo && Foo = bar
+    def visit_constant_and_write_node(node)
+      AndNode.new(
+        ConstantReadNode.new(node.name_loc),
+        ConstantWriteNode.new(node.name_loc, node.value, node.operator_loc, node.location),
+        node.operator_loc,
+        node.location
+      )
+    end
+
+    # Foo ||= bar
+    #
+    # becomes
+    #
+    # Foo || Foo = bar
+    def visit_constant_or_write_node(node)
+      OrNode.new(
+        ConstantReadNode.new(node.name_loc),
+        ConstantWriteNode.new(node.name_loc, node.value, node.operator_loc, node.location),
+        node.operator_loc,
+        node.location
+      )
+    end
+
+    # Foo += bar
+    #
+    # becomes
+    #
+    # Foo = Foo + bar
+    def visit_constant_operator_write_node(node)
+      desugar_operator_write_node(node, ConstantWriteNode, ConstantReadNode)
+    end
+
+    # Foo::Bar &&= baz
+    #
+    # becomes
+    #
+    # Foo::Bar && Foo::Bar = baz
```

Comment:
I think this desugaring is incorrect, because `Foo` should only be executed once, similar to calls.
i.e. for
```ruby
module Foo; end

(p Foo)::Bar ||= 42
```
Foo is printed once and not twice by CRuby.

And I guess `(p Foo)::Bar ||= 42` is a `ConstantPathAndWriteNode` (can't check right now).
But even if it's not it would be observable e.g. if `Foo` triggers `const_missing`.

I think the simplest fix for now is to not desugar the `ConstantPath*WriteNode` since that's invalid semantically, but still desugar the rest where it's valid.

Reply:
I'll make a PR for this.

Reply:
Thanks that would be great

Reply:
As expected:
```ruby
module Foo; end
# original
(p Foo)::Bar ||= 42 # prints Foo once
# vs
# desugared
defined?((p Foo)::Bar) ? (p Foo)::Bar : (p Foo)::Bar = 42 # prints Foo twice
```

A more complicated edge case:
```ruby
module FooM; end
class Object
  def self.const_missing(name)
    p name
    name == :Foo ? FooM : super
  end
end

# original
Foo::Bar ||= 42
Foo::Bar ||= 43
p Foo::Bar # => 42

# vs
# desugared
defined?(Foo::Bar) ? Foo::Bar : Foo::Bar = 42
defined?(Foo::Bar) ? Foo::Bar : Foo::Bar = 43
p Foo::Bar # => 43, and warning: already initialized constant FooM::Bar
```
which I think is because in reality it's like:
```ruby
tmp = Foo
defined?(tmp::Bar) ? tmp::Bar : tmp::Bar = value
```
and `defined?(Foo::Bar)` is just `nil` instead, even if `FooM::Bar` is set.